src/lab1-tests.md

```diff
@@ -14,91 +14,90 @@ Here's the testing boilerplate.
 
-Testing `show`
---------------
-
-The `show` function is supposed to print some HTML to the screen, but
-skip the tags inside.
-
-    >>> lab1.show('<body>hello</body>')
-    hello
-    >>> lab1.show('he<body>llo</body>')
-    hello
-    >>> lab1.show('he<body>l</body>lo')
-    hello
-    >>> lab1.show('he<body>l<div>l</div>o</body>')
-    hello
-
-Note that the tags do not have to match:
-
-    >>> lab1.show('he<body>l</div>lo')
-    hello
-    >>> lab1.show('he<body>l<div>l</body>o</div>')
-    hello
-
-Testing `URL`
--------------
+1.4 Telnet in Python
+--------------------
 
 Constructing a `URL` object parses a URL:
 
-    >>> lab1.URL('http://test.test/example1')
-    URL(scheme=http, host=test.test, port=80, path='/example1')
+    >>> lab1.URL('http://test.test/example1') #doctest: +ELLIPSIS
+    URL(scheme=http, host=test.test,... path='/example1')
 
 This works even if there is no path:
 
-    >>> lab1.URL('http://test.test')
-    URL(scheme=http, host=test.test, port=80, path='/')
-
-An explicit port will be used:
+    >>> lab1.URL('http://test.test') #doctest: +ELLIPSIS
+    URL(scheme=http, host=test.test,... path='/')
 
-    >>> lab1.URL('http://test.test:90')
-    URL(scheme=http, host=test.test, port=90, path='/')
-
-Testing `request`
------------------
+The `...` allow you to print a port once you implement that in Section
+1.7.
+    
+The first half of the `request` function is tested in the next section.
 
-The `request` function makes HTTP requests.
+1.5 Request and Response
+------------------------
 
-To test it, we use the `test.socket` object, which mocks the HTTP server:
+The `request` function makes HTTP requests. To test it, we use the
+`test.socket` object, which mocks the HTTP server:
 
-    >>> url = 'http://test.test/example1'
-    >>> test.socket.respond(url, b"HTTP/1.0 200 OK\r\n" +
-    ... b"Header1: Value1\r\n\r\n" +
-    ... b"Body text")
+    >>> url = test.socket.serve("Body text")
 
 Then we request the URL and test both request and response:
 
     >>> body = lab1.URL(url).request()
     >>> test.socket.last_request(url)
-    b'GET /example1 HTTP/1.0\r\nHost: test.test\r\n\r\n'
+    b'GET /0 HTTP/1.0\r\nHost: test\r\n\r\n'
     >>> body
     'Body text'
 
 With an unusual `Transfer-Encoding` the request should fail:
-
     
-    >>> url = 'http://test.test/te'
-    >>> test.socket.respond(url, b"HTTP/1.0 200 OK\r\n" +
-    ... b"Transfer-Encoding: chunked\r\n\r\n" +
-    ... b"0\r\n\r\n")
+    >>> url = test.socket.serve("", headers={
+    ...     "Transfer-Encoding": "chunked"
+    ... })
     >>> lab1.URL(url).request()
     Traceback (most recent call last):
       ...
     AssertionError
 
 Likewise with `Content-Encoding`:
     
-    >>> url = 'http://test.test/ce'
-    >>> test.socket.respond(url, b"HTTP/1.0 200 OK\r\n" +
-    ... b"Content-Encoding: gzip\r\n\r\n" +
-    ... b"\x00\r\n\r\n")
+    >>> url = test.socket.serve("", headers={
+    ...     "Content-Encoding": "chunked"
```

Comment:
```suggestion
    ...     "Content-Encoding": "gzip"
```

?

Reply:
Sure. Any value at all should be an error.